clang/lib/CIR/Dialect/IR/CIRAttrs.cpp

```diff
@@ -295,6 +295,61 @@ LogicalResult IntAttr::verify(function_ref<InFlightDiagnostic()> emitError,
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// FloatAttr definitions
+//===----------------------------------------------------------------------===//
+
+Attribute cir::FloatAttr::parse(AsmParser &parser, Type odsType) {
+  double value;
+
+  if (!odsType.isa<cir::FloatType>())
+    return {};
+  auto ty = odsType.cast<cir::FloatType>();
+
+  if (parser.parseLess())
+    return {};
+
+  if (parser.parseFloat(value))
+    parser.emitError(parser.getCurrentLocation(),
+                     "expected floating-point value");
+
+  if (parser.parseGreater())
+    return {};
+
+  auto losesInfo = false;
+  APFloat convertedValue{value};
+  convertedValue.convert(APFloat::EnumToSemantics(ty.getFloatSemantics()),
+                         llvm::RoundingMode::TowardZero, &losesInfo);
+
+  return cir::FloatAttr::get(ty, convertedValue);
+}
+
+void cir::FloatAttr::print(AsmPrinter &printer) const {
+  printer << '<' << getValue() << '>';
+}
+
+cir::FloatAttr cir::FloatAttr::getZero(mlir::cir::FloatType type) {
+  return get(type, APFloat::getZero(
+                       APFloat::EnumToSemantics(type.getFloatSemantics())));
```

Comment:
Have you considered `#cir.zero` for the float type? Zero gets more complicated with floating point so I wonder if there are reasons not to.

Reply:
Seems `#cir.zero` is indeed a better option. Although multiple object representations can be used for representing `0` for floating-point types, upstream clang generates '\0' bytes for floating-point `0`:

```cpp
float x = 0;
// x:
//         .long   0x00000000                      # float 0

double y = 0;
// y:
//         .quad   0x0000000000000000              # double 0
```

One nits here: a floating-point zero actually has two possible values: `+0` and `-0`. We may assume that `#cir.zero : !cir.float` always represents `+0`.

Reply:
> upstream clang generates '\0' bytes for floating-point 0

what's the LLVM output? 

> One nits here: a floating-point zero actually has two possible values: +0 and -0. We may assume that #cir.zero : !cir.float always represents +0.

Is that the same assumption that is made for what the object return by `APFloat::getZero` represents?

Reply:
Sounds good to me!